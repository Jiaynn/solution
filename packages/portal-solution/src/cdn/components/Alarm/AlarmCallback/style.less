.comp-alarm-callback {
  color: #333;

  .callback-config {
    &-label {
      margin-right: 16px;
      font-weight: 500;
    }

    &-uri {
      margin-right: 12px;
    }

    &-btn {
      margin-right: 8px;
    }
    // 覆盖全局的配置
    .ant-btn + .ant-btn {
      margin-left: 0;
    }
  }

  &-apply {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    margin-top: 120px;
    color: #333;

    .apply-title {
      font-size: 24px;
      font-weight: 500;
      color: #333;
    }

    .apply-submit,
    .apply-tips {
      margin-top: 48px;
      text-align: center;
    }
  }
}
