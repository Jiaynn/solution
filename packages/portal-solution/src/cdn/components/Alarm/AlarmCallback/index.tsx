/**
 * @file Alarm Callback Manage
 * @author linchen <gakiclin@gmail.com>
 */

import React from 'react'
import { observer } from 'mobx-react'
import { FieldState } from 'formstate-x'
import { useInjection } from 'qn-fe-core/di'
import Spin from 'react-icecream/lib/spin'
import Button from 'react-icecream/lib/button'
import Modal from 'react-icecream/lib/modal'
import Form from 'react-icecream/lib/form'
import Input from 'react-icecream/lib/input'
import { useFormstateX } from 'react-icecream-2/form-x'
import { ToasterStore as Toaster } from 'portal-base/common/toaster'
import { bindFormItem, bindTextInput } from 'portal-base/common/form'
import { Link } from 'portal-base/common/router'

import { textPattern } from 'cdn/transforms/form'

import { ModalStore, IModalProps } from 'cdn/stores/modal'

import { useAsync } from 'cdn/hooks/api'

import callbackDoc from 'cdn/docs/alarm-callback.pdf'

import { httpUrl } from 'cdn/constants/pattern'

import TipIcon from 'cdn/components/TipIcon'

import AlarmCallbackApis, { CallbackInfo } from 'cdn/apis/alarm/callback'

import './style.less'

interface ModalProps extends IModalProps {
  title?: string
  visible: boolean
  state: State
}

const CallbackUriModel = observer(function _CallbackUriModel({
  state,
  title,
  visible,
  onCancel,
  onSubmit
}: ModalProps) {

  const handleSubmit = (event: React.SyntheticEvent) => {
    event.preventDefault()
    state.validate().then(({ hasError }) => {
      if (!hasError) {
        onSubmit()
      }
    })
  }

  return (
    <Modal
      onOk={handleSubmit}
      onCancel={onCancel}
      visible={visible}
      title={title}
    >
      <Form.Item
        colon={false}
        label="告警回调地址"
        {...bindFormItem(state)}
        extra="请输入公网可访问的 URL 地址，设置后将对所有 CDN、DCDN 的域名告警配置生效。"
      >
        <Input {...bindTextInput(state)} />
      </Form.Item>
    </Modal>
  )
})

function createState(info: CallbackInfo) {
  return new FieldState(info.callbackUri || '').validators(
    v => textPattern(httpUrl)(v)
  )
}

type State = ReturnType<typeof createState>

function getValue(state: State) {
  return state.value
}

function useAlarmCallback() {
  const {
    isLoading,
    isIdle,
    isError,
    error,
    run,
    result
  } = useAsync<{ callbackUri?: string }>({ callbackUri: undefined })

  const alarmCallbackApis = useInjection(AlarmCallbackApis)

  const call = React.useCallback(() => (
    run(alarmCallbackApis.getAlarmCallback())
  ), [run, alarmCallbackApis])

  React.useEffect(() => { call() }, [call])

  const toaster = useInjection(Toaster)

  React.useEffect(() => {
    if (isError) {
      toaster.error('查询告警回调信息失败')
    }
  }, [toaster, error, isError])

  return {
    call,
    result,
    isIdle,
    isLoading
  }
}

export default observer(function AlarmCallback() {
  const { call, result, isIdle, isLoading } = useAlarmCallback()
  const alarmCallbackApis = useInjection(AlarmCallbackApis)

  const state = useFormstateX(createState, [result])

  const [modalStore] = React.useState(
    () => new ModalStore<{ title?: string }>()
  )

  const toaster = useInjection(Toaster)

  const doUpsert = React.useCallback((title: string) => {
    toaster.promise(
      modalStore.open({ title }).then(
        () => alarmCallbackApis.upsertAlarmCallback(getValue(state))
      ).then(call)
    )
  }, [alarmCallbackApis, modalStore, toaster, state, call])

  const doDelete = React.useCallback(() => {
    toaster.promise(
      alarmCallbackApis.deleteAlarmCallback().then(call)
    )
  }, [alarmCallbackApis, toaster, call])

  const handleDelete = React.useCallback(() => {
    Modal.confirm({
      title: '删除提示',
      content: '删除后将停止向回调地址发送告警信息，该操作将同时影响 CDN、DCDN 的告警通知回调配置，如需恢复请重新配置。',
      onOk: doDelete
    })
  }, [doDelete])

  const callbackUri = result && result.callbackUri

  const configCnt = callbackUri
    ? (
      <>
        <span className="callback-config-uri">{callbackUri}</span>
        <Button
          onClick={() => doUpsert('修改告警回调地址')}
          className="callback-config-btn"
          type="link"
        >修改</Button>
        <Button
          onClick={handleDelete}
          className="callback-config-btn"
          type="link"
        >删除</Button>
      </>
    )
    : (
      <>
        <span className="callback-config-uri">暂无配置</span>
        <Button
          onClick={() => doUpsert('添加告警回调地址')}
          className="callback-config-btn"
          type="link"
        >添加</Button>
      </>
    )

  const tipContent = (
    <>
      <Link target="_blank" to={callbackDoc} rel="noopener noreferrer">告警通知回调&nbsp;</Link>
      功能支持将已配置的域名告警信息回调到固定的公网 URL，帮助您将告警信息集成到已有运维系统或消息通知系统中。 回调地址支持 http 和 https 协议，配置后会对所有已配置的域名告警立即生效，请保证回调地址处于可访问的状态。
    </>
  )

  return (
    <div className="comp-alarm-callback">
      <Spin spinning={isIdle || isLoading}>
        <div className="callback-config">
          <span className="callback-config-label">
            告警回调地址
            <TipIcon tip={tipContent} />
          </span>
          {configCnt}
        </div>
        <CallbackUriModel
          state={state}
          {...modalStore.bind()}
        />
      </Spin>
    </div>
  )
})
