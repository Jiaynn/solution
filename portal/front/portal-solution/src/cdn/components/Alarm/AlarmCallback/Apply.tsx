/**
 * @file Alarm Callback Apply
 * @author linchen <gakiclin@gmail.com>
 */

import React from 'react'

import Button from 'react-icecream/lib/button'

import './style.less'

export default function AlarmCallbackApply() {
  return (
    <div className="comp-alarm-callback-apply">
      <h4 className="apply-title">暂未开通</h4>
      <p className="apply-tips">
        告警通知回调功能支持将已配置的域名告警信息回调到固定的公网 URL，帮助您将告警信息集成到已有运维系统或消息通知系统中。
        <br />
        该功能目前在内测阶段，如需开通可以提交工单申请。
      </p>
      <Button
        target="_blank"
        href="https://support.qiniu.com/tickets/new/form?category=%E9%85%8D%E7%BD%AE%E9%97%AE%E9%A2%98&space=CDN"
        className="apply-submit"
        type="primary"
      >提交工单</Button>
    </div>
  )
}
