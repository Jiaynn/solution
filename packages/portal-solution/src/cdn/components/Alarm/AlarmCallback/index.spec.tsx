import React from 'react'

import { RendererUtils as Renderer } from 'test'

import AlarmCallback from '.'

it('renders correctly', () => {
  const renderer = new Renderer()
  const tree = renderer.createWithAct(
    <AlarmCallback />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

