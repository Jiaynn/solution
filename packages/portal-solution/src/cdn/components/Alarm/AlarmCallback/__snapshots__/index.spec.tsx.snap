// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="comp-alarm-callback"
>
  <div
    className="ant-spin-nested-loading"
  >
    <div>
      <div
        className="ant-spin ant-spin-spinning"
      >
        <span
          className="ant-spin-dot ant-spin-dot-spin"
        >
          <i
            className="ant-spin-dot-item"
          />
          <i
            className="ant-spin-dot-item"
          />
          <i
            className="ant-spin-dot-item"
          />
          <i
            className="ant-spin-dot-item"
          />
        </span>
      </div>
    </div>
    <div
      className="ant-spin-container ant-spin-blur"
    >
      <div
        className="callback-config"
      >
        <span
          className="callback-config-label"
        >
          告警回调地址
          <i
            aria-label="icon: info-circle"
            className="anticon anticon-info-circle"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "color": "#999",
                "cursor": "pointer",
                "fontSize": "14px",
                "fontWeight": "normal",
                "marginLeft": "4px",
                "verticalAlign": "-0.125em",
              }
            }
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="info-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
              />
              <path
                d="M464 336a48 48 0 1 0 96 0 48 48 0 1 0-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
              />
            </svg>
          </i>
        </span>
        <span
          className="callback-config-uri"
        >
          暂无配置
        </span>
        <button
          className="ant-btn callback-config-btn ant-btn-link"
          onClick={[Function]}
          type="button"
        >
          <span>
            添加
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
